### PR TITLE
Fold ipv4ll fallback modes into normal ipv4ll addressing

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1120,7 +1119,7 @@ static int link_acquire_ipv4_conf(Link *link) {
         assert(link->manager);
         assert(link->manager->event);
 
-        if (link_ipv4ll_enabled(link, ADDRESS_FAMILY_IPV4)) {
+        if (link_ipv4ll_enabled(link)) {
                 assert(link->ipv4ll);
```
Comment (yuwata) [inline src/network/networkd-link.c:1130]: I prefer to simply use `if (link->ipv4ll)`.

File: src/network/networkd-dhcp4.c [hunk 2]
```diff
@@ -1088,6 +1088,21 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
                                         return -ENOMSG;
                         }
                         break;
+
+                case SD_DHCP_CLIENT_EVENT_TRANSIENT_FAILURE:
+                        if (link_ipv4ll_enabled(link)) {
+                                assert(link->ipv4ll);
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c:1094]: Same here. I prefer `if (link->ipv4ll)`

**Review by yuwata — COMMENTED** [review]
I've not tested this. I may missed something, but it seems there is no logic that dropping ipv4ll after DHCPv4 address is acquired when once DHCPv4 is failed and ipv4ll is acquired:
1. DHCPv4 fails (3 trial),
2. IPv4LL is acquired,
3. DHCPv4 address is acquired,
4. dropping IPv4LL <--- is this implemented in this PR?

**Comment by yuwata** [comment]
Replaced by #17692.

**Comment by BillPlunkett** [comment]
I've been following this discussion for a while, but not 100% sure I understand where things stand.

The desired behavior (for me) is:
1. network configured in DHCP client mode
2. DHCP server not present or unresponsive, so ipv4ll address assigned
3. DHCP server (re)appears, so ipv4ll address removed and DHCP address assigned

From above, it seems like this behavior has been implemented in v247 with the exception that the ipv4ll address will remain after a DHCP address is assigned in step 3.  Assuming that is true, what are the ramifications?

**Comment by keszybz** [comment]
Let's continue the discussion on the other PR.

**Pull request closed without merging**